monai/losses/image_dissimilarity.py

```diff
@@ -0,0 +1,153 @@
+
+    Adapted from:
+        https://github.com/voxelmorph/voxelmorph/blob/legacy/src/losses.py
+        DeepReg (https://github.com/DeepRegNet/DeepReg)
+    """
+
+    def __init__(
+        self,
+        in_channels: int,
+        ndim: int = 3,
+        kernel_size: int = 9,
+        kernel_type: str = "rectangular",
+        reduction: Union[LossReduction, str] = LossReduction.MEAN,
+    ) -> None:
+        """
+        Args:
+            in_channels: number of input channels
+            ndim: number of spatial ndimensions, {``1``, ``2``, ``3``}. Defaults to 3.
+            kernel_size: kernel size or kernel sigma for kernel_type=``"gaussian"``
+            kernel_type: {``"rectangular"``, ``"triangular"``, ``"gaussian"``}. Defaults to ``"rectangular"``.
+            reduction: {``"none"``, ``"mean"``, ``"sum"``}
+                Specifies the reduction to apply to the output. Defaults to ``"mean"``.
+
+                - ``"none"``: no reduction will be applied.
+                - ``"mean"``: the sum of the output will be divided by the number of elements in the output.
+                - ``"sum"``: the output will be summed.
+        """
+        super(LocalNormalizedCrossCorrelationLoss, self).__init__(reduction=LossReduction(reduction).value)
+        self.in_channels = in_channels
+        self.ndim = ndim
+        if self.ndim not in [1, 2, 3]:
+            raise ValueError(f"Unsupported ndim: {self.ndim}-d, only 1-d, 2-d, and 3-d inputs are supported")
+        self.kernel_size = kernel_size
+        if kernel_type == "rectangular":
+            self.kernel, self.kernel_vol, self.padding = self.make_rectangular_kernel()
+        elif kernel_type == "triangular":
+            self.kernel, self.kernel_vol, self.padding = self.make_triangular_kernel()
+        elif kernel_type == "gaussian":
+            self.kernel, self.kernel_vol, self.padding = self.make_gaussian_kernel()
+        else:
+            raise ValueError(
+                f'Unsupported kernel_type: {kernel_type}, available options are ["rectangular", "triangular", "gaussian"].'
+            )
+
+    def make_rectangular_kernel(self):
+        shape = [1, self.in_channels] + [self.kernel_size] * self.ndim
+        return torch.ones(shape, dtype=torch.float), self.kernel_size ** self.ndim, int((self.kernel_size - 1) / 2)
+
+    def make_triangular_kernel(self):
+        fsize = torch.tensor((self.kernel_size + 1) / 2, dtype=torch.int)
+        f1 = torch.ones([1, 1] + [fsize] * self.ndim, dtype=torch.float) / fsize  # (1, 1, D, H, W)
+        f2 = (
+            torch.ones([self.in_channels, 1] + [fsize] * self.ndim, dtype=torch.float) / fsize
+        )  # (1, in_channels, D, H, W)
+        # (1, 1, D, H, W) -> (1, in_channels, D, H, W)
+        fn = conv_dict[self.ndim]
+        kernel = fn(f1, f2, padding=int((fsize - 1) / 2))
+
+        return kernel, torch.sum(kernel ** 2), int((fsize - 1) / 2)
+
+    def make_gaussian_kernel(self):
```

Comment:
just wanted to mention this in case it's useful: https://github.com/Project-MONAI/MONAI/blob/db7bf8984712f3e46ed533611a70c9b8cd0293bd/monai/networks/layers/convutils.py#L81-L85

Reply:
Thanks for mentioning!
Actually, the DeepReg implementation of gaussian kernel uses `(self.kernel_size - 1) / 2.0` as the channel-wise mean out of unknown reason. 
Do we want to correct this by using MONAI's `gaussian_1d` or stick with DeepReg's implementation so we can compare results with DeepReg for unit test?

Reply:
thanks, perhaps we keep the deepreg one for now since we know it works. we could always improve the impl. with another follow-up PR

Reply:
> Thanks for mentioning!
> Actually, the DeepReg implementation of gaussian kernel uses `(self.kernel_size - 1) / 2.0` as the channel-wise mean out of unknown reason.
> Do we want to correct this by using MONAI's `gaussian_1d` or stick with DeepReg's implementation so we can compare results with DeepReg for unit test?
> @YipengHu

They uses different origins to sample probability densities, should be equivalent (if implemented correctly).

Reply:
> > Thanks for mentioning!
> > Actually, the DeepReg implementation of gaussian kernel uses `(self.kernel_size - 1) / 2.0` as the channel-wise mean out of unknown reason.
> > Do we want to correct this by using MONAI's `gaussian_1d` or stick with DeepReg's implementation so we can compare results with DeepReg for unit test?
> > @YipengHu
> 
> They use different origins to sample probability densities, should be equivalent (if implemented correctly).

The monai `gaussian_1d` enforces centre distribution (the centre pixel of the kernel to be 0), while in DeepReg, the kernel is not symmetric along the channel dimension.
Do you think it will actually be better to enforce centre distribution on the channel dimension?